### PR TITLE
[SE-0326] Re-enable multi-statement closure inference by default

### PR DESCRIPTION
**Pull request by xedin**
Fixes a performance regression related to `one-way` constraints and re-enables
multi-statement closure inference by default.

The changes include:

- New locator for each pattern binding element
- Recording a type of a named pattern declaration in a constraint system
- Solving pattern bindings declarations via conjunctions

Resolves: rdar://88205521

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
@swift-ci please test source compatibility

File: lib/Sema/CSGen.cpp [hunk 1]
```diff
@@ -2299,9 +2299,71 @@ namespace {
...
+          // `$T_x_y = ($T_s_x, $T_s_y)` where both `$T_s_x` and `$T_s_y` have
+          // to allow l-value, but `$T_x_y` does not. Early simplication of `=`
+          // constraint (due to right-hand side being a "concrete" tuple type)
+          // would drop l-value option from `$T_s_x` and `$T_s_y` which leads to
+          // a failure during member lookup because `x` and `y` are both
+          // `@lvalue Int`. To avoid that, declaration type would mimic pattern
+          // type with all l-value options stripped, so the equality constraint
+          // becomes `($T_x, $_T_y) = ($T_s_x, $T_s_y)` which doesn't result in
+          // stripping of l-value flag from the right-hand side since
+          // simplication can only happen when either side is resolved.
```
Comment (LucianoPAlmeida) [inline lib/Sema/CSGen.cpp:2339]: Typo "simplication" -> "simplification"? =]

Comment (xedin) [reply lib/Sema/CSGen.cpp:2339]: 🤦‍♂️ I'm going to fix it after the source suite is done...

**Comment by xedin** [comment]
There are a couple of unrelated issues in the source compatibility suite:

- Eureka is crashing in request evaluator:
```
Cannot build interface type for term [OptionsProviderRow:OptionsProviderType].[OptionsProviderConformance:Option]
Prefix does not conform to any protocols: [OptionsProviderRow:OptionsProviderType]
```

- Moya conformance failure:

```
error: type 'MoyaProvider<Target>' does not conform to protocol 'ReactiveCompatible'
extension MoyaProvider: ReactiveCompatible {}
```

- SwiftGraph conformance failure due to unavailable subscript:

```
error: type 'Graph<V>' does not conform to protocol 'Collection'
open class Graph<V: Equatable>: CustomStringConvertible, Sequence, Collection {
           ^
/Users/ec2-user/jenkins/workspace/swift-PR-source-compat-suite-debug-macos/swift-source-compat-suite/project_cache/SwiftGraph/Sources/SwiftGraph/Graph.swift:22:12: error: unavailable subscript 'subscript(_:)' was used to satisfy a requirement of protocol 'Collection'
open class Graph<V: Equatable>: CustomStringConvertible, Sequence, Collection {
           ^
```

File: lib/Sema/CSGen.cpp (hunk 1 above)
Comment (xedin) [reply lib/Sema/CSGen.cpp]: Fixed!

**Comment by shahmishal** [comment]
@swift-ci please test source compatibility

**Comment by shahmishal** [comment]
> @shahmishal FYI source compat suite is failing with `[/Users/ec2-user/jenkins/workspace-private/swift-PR-source-compat-suite-macos/swift/utils/build-script] ERROR: can't find CMake (please install CMake)`

This should be fixed now.

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
@swift-ci please test

**Review by hborla — APPROVED** [review]
Awesome, lgtm!

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test Linux platform

**Comment by xedin** [comment]
@swift-ci please test macOS platform

**Pull request merged**